src/charm_relation_building_blocks/relation_handlers/application_data.py

```diff
@@ -0,0 +1,187 @@
+                           By default, this charm already observes the relation_changed event.
+        """
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+        self._schema = data_model
+
+        if not refresh_event:
+            refresh_event = []
+        if isinstance(refresh_event, BoundEvent):
+            refresh_event = [refresh_event]
+        for ev in refresh_event:
+            self.framework.observe(ev, self.on_relation_changed)
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed, self.on_relation_changed
+        )
+
+    def __len__(self):
+        """Return the number of related applications."""
+        return len(self.get_relations())
+
+    def on_relation_changed(self, _: EventBase) -> None:
+        """Handle when the remote application data changed."""
+        self.on.data_changed.emit()
+
+    def get_relations(self):
+        """Return the relation instances for applications related to us on the monitored relation."""
+        return self._charm.model.relations.get(self._relation_name, ())
+
+    def get_data(self) -> Optional[DatabagModel]:
+        """Return data for at most one related application, raising if more than one is available.
+
+        Useful for charms that always expect exactly one related application.  It is recommended that those charms also
+        set limit=1 for that relation in charmcraft.yaml.  Returns None if no data is available (either because no
+        applications are related to us, or because the related application has not sent data).
+        """
+        relations = self.get_relations()
+        if len(relations) == 0:
+            return None
+        if len(relations) > 1:
+            # TODO: Different exception type?
+            raise ValueError("Cannot get_info when more than one application is related.")
+
+        raw_data = relations[0].data.get(relations[0].app)
+        if not raw_data:
+            return None
+
+        # Static analysis errors saying the keys may not be strings.  Protect against this by converting them.
+        raw_data = {str(k): v for k, v in raw_data.items()}
+
+        return self._schema(**raw_data)
+
+    def get_data_from_all_relations(self) -> List[DatabagModel]:
+        """Return a list of data objects from all relations."""
+        relations = self.get_relations()
+        info_list = []
+        for i, relation in enumerate(relations):
+            data_dict = relation.data.get(relation.app)
+            if not data_dict:
+                info_list.append(None)
+                continue
+
+            # Static analysis errors saying the keys may not be strings.  Protect against this by converting them.
+            data_dict = {str(k): v for k, v in data_dict.items()}
+            info_list.append(self._schema(**data_dict))
+        return info_list
```

Comment:
Continuing discussions from [here](https://github.com/canonical/istio-k8s-operator/pull/44#discussion_r1938324451).  This function feels like its a useful tool, but awkwardly named.  Suggestions appreciated